EHR retrieval and query with GenAI models (#504)
nvidia-holoscan/holohub: applications/ehr_query_llm/fhir/README.md

This application also uses ZeroMQ to communicate with its own clients, listening on a well known port on local host for messages to retrieve resources of a patient, as well as publishing the retrieved resources on another well known port. For simplicity, the listening port is defined in the code to be `5600`, and the publishing port `5601`. Messaging security, at transport or message level, is not implemented in this example.

Message schema is simple, with a well known topic string and topic specific content schema in JSON format.

The default set of FHIR resource types to retrieve are listed below, which can be overridden by the request message
- Observation
- ImagingStudy
- FamilyMemberHistory
- Condition
- DiagnosticReport
- DocumentReference

## Requirements

- On a [Holohub supported platform](../../README.md#supported-platforms)
- Python 3.10+
- Python packages on [Pypi](https://pypi.org), including holoscan, fhir.resources, holoscan, pyzmq, requests and their dependencies

## Run Instructions

> There are a number of ways to build and run this application, as well as packaging this application as a Holoscan Application Package, a [Open Container Initiative](https://opencontainers.org/) compliant image. The following sections describe each in detail.

Reviewer: ```suggestion
There are several ways to build and run this application and package it as a Holoscan Application Package, an [Open Container Initiative](https://opencontainers.org/) compliant image. The following sections describe each in detail.
```

Author: Corrected per review comments

It is further expected that you have read the [HoloHub README](../../../README.md), have cloned the HoloHub repository to your local system, and the current working directory is the HoloHub root, `holohub`.

**_Note_**:

> The application listens on request message to start retrieving resources from the server and then publishes the results, so another application is needed to drive this workflow, e.g. the LLM applicaition. To help with simple testing, a Python script is provided as part of this application, and its usage is described below in this [section](#test-the-running-application).

Author: Thanks for catching this! It had been proven a few times that `codespell` misses typos, while giving false positives on acronyms.

      self._acquire()
    File "/workspace/holohub/.local/lib/python3.10/site-packages/filelock/_unix.py", line 42, in _acquire
      fd = os.open(self.lock_file, open_flags, self._context.mode)
  PermissionError: [Errno 13] Permission denied: '/usr/local/lib/python3.10/dist-packages/holoscan-2.4.0.dist-info/axle.lck'
```

### Run the Application in Dev Container

This is a step wise way to run the application in a dev container.
```bash
./dev_container build --docker_file applications/ehr_query_llm/fhir/Dockerfile --img holoscan:fhir --verbose --no-cache
```

Optionally check the newly built image
```bash
$ docker images
REPOSITORY         TAG               IMAGE ID       CREATED          SIZE
holoscan           fhir              508140b8d446   3 minutes ago    14.1GB
```

Launch the container
```bash
./dev_container launch --img holoscan:fhir --as_root
```

Now in the container, build and run the application

```bash
root:~# pwd
/workspace/holohub

root:~# ./run clear_cache
root:~# ./run build fhir
root:~# ./run launch fhir --extra_args "--fhir_url <f_url> --auth_url <a_url> --uid <id> --secret <token>"
```

Once done, `exit` the container.

**_Note_**:

> For a Python application, building and launching the dev container can be completely skipped, which means one can simply `pip` install the application specific dependencies in the `requirements.txt` file, and then run the few `./run` commands, shown above, to start the application directly in the host dev environment.

Author: > Can we remove this note in favor of the heading below?

Removed per review comment.

### Run the Application in the Host Dev Environment with dev_container script

First create and activate a Python virtual environment, followed with installing the dependencies

```bash
python3 -m venv .testenv
source .testenv/bin/activate
pip install -r applications/ehr_query_llm/fhir/requirements.txt
```

Build and install the application with `dev_container`
```bash
./dev_container build_and_install fhir
```

Now, run the application which is _installed_ in the `install` folder, with server URLs and credential of your own
```bash
python install/bin/fhir/python/ --fhir_url <f_url> --auth_url <a_url> --uid <id> --secret <token>

> ```

Author: Good point! I was also trying to work around a limitation on the HSDK CLI where currently it only packages modules and such under a single root folder, and `install` folder with the help of the CMake file allowed me to stage all files including libs and operators under the same `install/bin/<app>/python` folder.

Once enhanced CLI with the `--add` option is in for HSDK v2.7, I can better separate the concerns for running HoloHub container and packaging Holoscan App.

Will see if I can make the current setup better, in any case.
